### ISSUE TITLE
`Cannot inherit frozen dataclass from a non-frozen one` error when using pydantic `frozen=True`

### ISSUE THREAD

**Issue by johnnyleitrim** [python/mypy#14857]
**Bug Report**

Specifying `frozen=True` as model class kwargs for pydantic models causes the following error to be emitted from `mypy`:
```
Cannot inherit frozen dataclass from a non-frozen one
```

**To Reproduce**
1. Install pydantic and mypy: `pip install pydantic mypy`
2. Create `test.py` with the following content:
```python
from pydantic import BaseModel

class MyModel(BaseModel, frozen=True):
    my_val: int
```
3. Run `mypy` on the test file: `mypy test.py`

**Expected Behavior**
No errors reported (as was the case with mypy 1.0.1):
```
> mypy test.py
Success: no issues found in 1 source file
```

**Actual Behavior**
Errors reported:
```
> mypy test.py
test.py:3: error: Cannot inherit frozen dataclass from a non-frozen one  [misc]
Found 1 error in 1 file (checked 1 source file)
```

**Your Environment**

- Mypy version used: 1.1.1
- Python version used: 3.10

**Comment by wesleywright** [comment]
Similar to in #14850, I believe this is actually an issue with the Pydantic library's use of `dataclass_transform`. From looking at the [code](https://github.com/pydantic/pydantic/blob/main/pydantic/main.py#L147), it seems that `BaseModel` is itself using the transform and should be treated as a dataclass, so per the standard, frozen dataclasses cannot subclass it.

You may be able to use the same workaround as suggested in #14850 (using the latest version of the Pydantic plugin).

**Comment by wesleywright** [comment]
After double checking the spec (whoops), there is actually a [provision](https://peps.python.org/pep-0681/#dataclass-semantics) for this:

> Frozen dataclasses cannot inherit from non-frozen dataclasses. A class that has been decorated with dataclass_transform is considered neither frozen nor non-frozen, thus allowing frozen classes to inherit from it. Similarly, a class that directly specifies a metaclass that is decorated with dataclass_transform is considered neither frozen nor non-frozen.

This seems to be the only way to declare a class that has dataclass semantics but which is treated as neither frozen nor non-frozen. This is definitely a bug in MyPy; I will try to create a PR to fix it.
